File: apps/api/internal/auth/claims.go
package auth

import (
	"encoding/json"
	"strings"
)

// ClaimNames maps portfolio tenant fields onto IdP-specific claim keys (Auth0 / Entra / P01).
type ClaimNames struct {
	OrgID         string // default org_id
	Organizations string // default organizations (array of objects or strings)
}

func DefaultClaimNames() ClaimNames {
	return ClaimNames{OrgID: "org_id", Organizations: "organizations"}
}

func ParseClaimNames(orgClaim, orgsClaim string) ClaimNames {
	c := DefaultClaimNames()
	if v := strings.TrimSpace(orgClaim); v != "" {
		c.OrgID = v
	}
	if v := strings.TrimSpace(orgsClaim); v != "" {
		c.Organizations = v
	}
	return c
}

// claimString reads a string claim; supports dotted paths (e.g. extension_OrgId is a single key).
func claimString(claims map[string]any, key string) string {
	if key == "" || claims == nil {
		return ""
	}
	// Prefer exact key; also try first segment of comma-separated alternates.
	for _, k := range strings.Split(key, ",") {
		k = strings.TrimSpace(k)
		if k == "" {
			continue
		}
		if v, ok := claims[k]; ok {
			switch t := v.(type) {
			case string:
				return strings.TrimSpace(t)
			case json.Number:
				return strings.TrimSpace(t.String())
			case float64:
				return strings.TrimSpace(strings.TrimSuffix(strings.TrimSpace(jsonNumber(t)), ".0"))
			}
		}
	}
	return ""
}

func jsonNumber(f float64) string {
	b, _ := json.Marshal(f)
	return string(b)
}

func orgIDFromClaims(claims map[string]any, names ClaimNames) string {
	if id := claimString(claims, names.OrgID); id != "" {
		return id
	}
	// Fallbacks used by common IdPs when custom mapping is not set.
	for _, k := range []string{"org_id", "orgId", "organization_id", "tid"} {
		if id := claimString(claims, k); id != "" {
			return id
		}
	}
	orgs := organizationsFromClaims(claims, names)
	if len(orgs) > 0 {
		return orgs[0].OrgID
	}
	return ""
}

type orgMembership struct {
	OrgID   string
	OrgName string
	Role    string
}

func organizationsFromClaims(claims map[string]any, names ClaimNames) []orgMembership {
	if claims == nil {
		return nil
	}
	keys := []string{names.Organizations, "organizations", "orgs"}
	var raw any
	for _, k := range keys {
		if k == "" {
			continue
		}
		if v, ok := claims[k]; ok {
			raw = v
			break
		}
	}
	if raw == nil {
		return nil
	}
	arr, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]orgMembership, 0, len(arr))
	for _, item := range arr {
		switch t := item.(type) {
		case string:
			id := strings.TrimSpace(t)
			if id != "" {
				out = append(out, orgMembership{OrgID: id, OrgName: id, Role: "member"})
			}
		case map[string]any:
			id := firstString(t, "org_id", "orgId", "id")
			if id == "" {
				continue
			}
			out = append(out, orgMembership{
				OrgID:   id,
				OrgName: firstString(t, "org_name", "orgName", "name", "org_id", "orgId"),
				Role:    firstString(t, "role", "roles"),
			})
		}
	}
	return out
}

func firstString(m map[string]any, keys ...string) string {
	for _, k := range keys {
		if s := claimString(m, k); s != "" {
			return s
		}
	}
	return ""
}

File: apps/web/lib/oidc/claims.ts
/** Claim key mapping for BYO IdPs (Auth0 / Entra / P01). Comma-separated alternates allowed. */

export function orgClaimKeys(): string[] {
  const raw = process.env.OIDC_ORG_CLAIM?.trim() || "org_id,orgId,organization_id,tid";
  return raw.split(",").map((s) => s.trim()).filter(Boolean);
}

export function orgsClaimKeys(): string[] {
  const raw = process.env.OIDC_ORGS_CLAIM?.trim() || "organizations,orgs,org_memberships";
  return raw.split(",").map((s) => s.trim()).filter(Boolean);
}

export type OrgMembership = {
  orgId: string;
  orgName: string;
  role: string;
};

function asRecord(v: unknown): Record<string, unknown> | null {
  return v && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : null;
}

export function pickOrgId(claims: Record<string, unknown>): string | undefined {
  for (const key of orgClaimKeys()) {
    const v = claims[key];
    if (typeof v === "string" && v.trim()) return v.trim();
  }
  const orgs = pickOrganizations(claims);
  return orgs[0]?.orgId;
}

export function pickOrganizations(claims: Record<string, unknown>): OrgMembership[] {
  let raw: unknown;
  for (const key of orgsClaimKeys()) {
    if (claims[key] !== undefined) {
      raw = claims[key];
      break;
    }
  }
  if (!Array.isArray(raw)) return [];
  const out: OrgMembership[] = [];
  for (const item of raw) {
    if (typeof item === "string" && item.trim()) {
      out.push({ orgId: item.trim(), orgName: item.trim(), role: "member" });
      continue;
    }
    const o = asRecord(item);
    if (!o) continue;
    const orgId = [o.org_id, o.orgId, o.id].find((x) => typeof x === "string" && String(x).trim()) as
      | string
      | undefined;
    if (!orgId?.trim()) continue;
    const orgName = ([o.org_name, o.orgName, o.name, orgId].find(
      (x) => typeof x === "string" && String(x).trim(),
    ) || orgId) as string;
    const role = (typeof o.role === "string" && o.role) || "member";
    out.push({ orgId: orgId.trim(), orgName: String(orgName).trim(), role });
  }
  return out;
}
